modules/docs/arrow-docs/docs/docs/arrow/typeclasses/traverse/README.md

```diff
@@ -11,16 +11,491 @@ redirect_from:
 {:.intermediate}
 intermediate
 
-TODO. Meanwhile you can find a short description in the [intro to typeclasses]({{ '/docs/typeclasses/intro/' | relative_url }}).
+In functional programming it is very common to encode "behaviors" as data types - common behaviors include `Option` for possibly missing values, `Either` and `Validated` for possible errors, and [`Promise`]({{ '/docs/effects/promise/' | relative_url }}) for asynchronous computations, which has the same purpose as a `Future`, only pure and lazy.
 
+These behaviors tend to show up in functions working on a single piece of data - for instance parsing a single `String` into an `Int`, validating a login, or asynchronously fetching website information for a user.
+
+```kotlin:ank
+import arrow.core.Either
+import arrow.core.Option
+import arrow.fx.ForIO
+import arrow.fx.Promise
+
+interface Profile
+interface User
+
+fun userInfo(u: User): Promise<ForIO, Profile> = TODO()
+```
+
+Each function asks only for the data it actually needs; in the case of `userInfo`, a single `User`. We certainly could write one that takes a `List<User>` and fetch profile for all of them, would be a bit strange. If we just wanted to fetch the profile of just one user, we would either have to wrap it in a `List` or write a separate function that takes in a single user anyways. More fundamentally, functional programming is about building lots of small, independent pieces and composing them to make larger and larger pieces - does this hold true in this case?
```

Comment:
```suggestion
Each function asks only for the data it actually needs; in the case of `userInfo`, a single `User`. We could certainly write one which takes a `List<User>` and fetches a profile for all of them, but it would be a bit strange. If we wanted to fetch the profile of just one user, we would either have to wrap it in a `List` or write a separate function that takes in a single user anyways. More fundamentally, functional programming is about building lots of small, independent pieces and composing them to make larger and larger pieces - does this hold true in this case?
```

Reply:
thx